src/software/thunderscope/robot_communication.py

```diff
@@ -219,53 +238,68 @@ def __run_primitive_set(self) -> None:
 
-            # get the manual control primitive
-            diagnostics_primitive = DirectControlPrimitive(
-                motor_control=self.motor_control_diagnostics_buffer.get(block=False),
-                power_control=self.power_control_diagnostics_buffer.get(block=False),
+            # get the most recent fullsystem primitives
+            fullsystem_primitive_set = self.fullsystem_primitive_set_buffer.get(
+                block=True, timeout=ROBOT_COMMUNICATIONS_TIMEOUT_S
             )
 
-            # diagnostics is running
-            if self.robots_connected_to_manual:
-                # for all robots connected to diagnostics, set their primitive
-                for robot_id in self.robots_connected_to_manual:
-                    robot_primitives[robot_id] = Primitive(
-                        direct_control=diagnostics_primitive
-                    )
+            # filter for fullsystem controlled robots
+            fullsystem_robots = list(
+                robot_id
+                for robot_id, mode in self.robot_control_mode_map.items()
+                if mode == IndividualRobotMode.AI
+            )
+
+            # set fullsystem primitives for fullsystem robots
+            for robot_id in fullsystem_robots:
+                robot_primitives_map[robot_id] = (
+                    fullsystem_primitive_set.robot_primitives[robot_id]
+                )
 
             # sends a final stop primitive to all disconnected robots and removes them from list
             # in order to prevent robots acting on cached old primitives
-            for robot_id, num_times_to_stop in self.robots_to_be_disconnected.items():
+            for (
+                robot_id,
+                num_times_to_stop,
+            ) in self.robot_stop_primitive_send_count_map.items():
                 if num_times_to_stop > 0:
-                    robot_primitives[robot_id] = Primitive(stop=StopPrimitive())
-                    self.robots_to_be_disconnected[robot_id] = num_times_to_stop - 1
+                    robot_primitives_map[robot_id] = Primitive(stop=StopPrimitive())
+                    self.robot_stop_primitive_send_count_map[robot_id] = (
+                        num_times_to_stop - 1
+                    )
 
             # initialize total primitive set and send it
             primitive_set = PrimitiveSet(
                 time_sent=Timestamp(epoch_timestamp_seconds=time.time()),
                 stay_away_from_ball=False,
-                robot_primitives=(
-                    robot_primitives
-                    if not self.should_send_stop
-                    else {
-                        robot_id: Primitive(stop=StopPrimitive())
-                        for robot_id in robot_primitives.keys()
-                    }
-                ),
+                robot_primitives=self.robot_stop_primitives_map
+                if self.should_send_stop
+                else robot_primitives_map,
```

Comment:
the above code sets `StopPrimitives` in the `robot_primitives_map`. So do we need to have this conditional evaluation based on `self.should_send_stop`? It seems that sending `robot_primitives_map` all the time would work?

Reply:
cant be bothered to figure out what's going on. hopefully will be cleaned up in #3174

Reply:
My understanding is that `should_send_stop` is triggered when estop is pressed. Once the stop primitives are sent, this flag is flipped back.

The stop primitives override seem like they're for when we switch control modes.

Definitely seems like these two code paths could be merged but that can be a later problem